Support for subjob macro sets

eeach opened the pull request:
Add detection of subjob and selection of macro sets based on
  both main job and subjob.

Change configuration to use the windower config library.

Reduce the size of the default configuration. Addon no longer
  crashes when searching for an absent key.

On branch macrochanger-update
Changes to be committed:
	deleted:    addons/macrochanger/data/settings.txt
	new file:   addons/macrochanger/data/settings.xml
	modified:   addons/macrochanger/macrochanger.lua

eeach committed:
Support for subjob macro sets

Add detection of subjob and selection of macro sets based on
  both main job and subjob.

Change configuration to use the windower config library.

Reduce the size of the default configuration. Addon no longer
  crashes when searching for an absent key.

On branch macrochanger-update
Changes to be committed:
	deleted:    addons/macrochanger/data/settings.txt
	new file:   addons/macrochanger/data/settings.xml
	modified:   addons/macrochanger/macrochanger.lua

z16 commented on addons/macrochanger/data/settings.xml:
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.1" ?>
```
You should remove this file, it will be auto-generated on the first run with the default settings. If you ever update the default settings, this file will not be updated and any new settings will not be copied over. Instead the config library's own mechanism will add new default keys automatically, so you should just use the existing mechanism.

z16 commented on addons/macrochanger/macrochanger.lua:
```diff
@@ -26,143 +26,85 @@
+function sleep (n)
+  local t0 = os.clock()
+  while os.clock() - t0 <= n do end
+end
```
We have a sleep function you can use: `coroutine.sleep(2)`. That way you don't have to busy wait.

z16 commented on addons/macrochanger/macrochanger.lua:
```diff
@@ -26,143 +26,85 @@
 
-windower.register_event('job change',function ()
+function job_change ()
 -- Could use the job ID passed into this function, but the addon would have to include the resources library
```
This is not your fault, since the comment indicates it has been here for a while, but you should definitely use the provided IDs in combination with our resources. It's very light if you only use jobs and much more efficient than calling `get_player`, especially twice. The results of this function should be cached when possible, so if you do decide to do it this way, save it in a local variable first. But like I said, best to use the provided arguments.

z16 commented on addons/macrochanger/macrochanger.lua:
```diff
@@ -26,143 +26,85 @@
+  local job  = ''
+  local book = ''
+  local page = ''
+  if settings.globaldisable == 0 then
```
This looks like a boolean setting, so you can use a boolean variable. Will also make it easier on users, since this is written to their user files, so they can use `true` and `false` instead of `1` and `0`.

z16 commented on addons/macrochanger/macrochanger.lua:
```diff
@@ -26,143 +26,85 @@
+  local sjob = windower.ffxi.get_player().sub_job
+  local job  = ''
+  local book = ''
+  local page = ''
```
In Lua it's enough to just declare local variables, you don't need to provide values that will always be overwritten. Also, all of these could be declared in inner scope, book and page even when they're used:
```lua
    local page = --[[...]]
    local book = --[[...]]
```

z16 commented on addons/macrochanger/macrochanger.lua:
```diff
@@ -26,143 +26,85 @@
+  local book = ''
+  local page = ''
+  if settings.globaldisable == 0 then
+    if mjob and sjob and settings.macros[(mjob..sjob):lower()] then
```
I would suggest using an underscore to divide the jobs in the settings file, i.e. `mjob .. '_' .. sjob`. That will make the settings file look better and job combinations easier to find, and we use similar syntax for other heavily used addons like *GearSwap*.

z16 commented on addons/macrochanger/macrochanger.lua:
```diff
@@ -26,143 +26,85 @@
+      war = {book = 1, page = 1},
+    },
+  }
+  settings = config.load(defaults)
```
Setting the default settings should happen on file scope, not on the login event. It will automatically reparse the settings when a new character logs in, so you don't need to do it manually.

z16 commented on addons/macrochanger/macrochanger.lua:
```diff
@@ -26,143 +26,85 @@
+    },
+  }
+  settings = config.load(defaults)
+  if name then
```
Is there ever a situation when `name` is `nil`? If so that's likely a bug, let us know if you've encountered that and under which circumstances, if possible.

z16 reviewed (changes requested):
Added a few remarks, most of it looks ok, but please use 4 space indentation throughout the file. Most of what I wrote are just suggestions to leverage existing functions which you may have not been aware of, but it's up to you.

There is one issue that's a bit unsettling, namely that this update is incompatible with previous configurations. That means as soon as we roll it out to users it will stop working for them until they copy the settings over. It's not a huge drawback, but something you should take into account. It can be as simple as informing them: If you don't read any settings other than the defaults, print a line to the log saying that they need to set it up in the settings file (preferably with a path to the file).

In the same vein, I would remove the default settings for `WAR`. I assume you added them to show an example of how it works. While that works, I'd suggest writing a short `ReadMe.md`, which you can then refer to. This will also be displayed when users click on "About" under the addon in the launcher menu.

One final note, this is also not your error, but since you're already changing it could you please adjust the copyright to conform to our standards? Replace `(c)` with `©` and add his (or his and your) name where it says `<your name>` and so on. Would appreciate it, but you don't have to, otherwise I'll just do it myself after we merge this.

eeach committed:
Set tabs to 4 spaces per style convention

On branch macrochanger-update
Changes to be committed:
	modified:   addons/macrochanger/macrochanger.lua

eeach committed:
Tighten scope of local variables in job_change()

Move local variables out of the function's scope and into the flow control block where they are used.

Drop the null string assignment to variables.

On branch macrochanger-update
Changes to be committed:
	modified:   addons/macrochanger/macrochanger.lua

eeach committed:
Use provided sleep function

Drop locally defined sleep function in favor of coroutine.sleep().

On branch macrochanger-update
Changes to be committed:
	modified:   addons/macrochanger/macrochanger.lua

eeach committed:
Update the copyright to conform to Windower project standards

Copy BSD license from
  http://dev.windower.net/doku.php?id=lua:writing_addons#licensing.

Add copyright notice with regard to present changes.

On branch macrochanger-update
Changes to be committed:
	modified:   addons/macrochanger/macrochanger.lua

eeach committed:
Use an underscore to divide the jobs in settings.xml

Update settings.xml to use [main]_[sub] to represnent a main/sub
  job combination.

On branch macrochanger-update
Changes to be committed:
	modified:   addons/macrochanger/macrochanger.lua

eeach committed:
Use the provided IDs and the resources library

Change job_change() so that it uses IDs provided by the job change
  Windower event and resolved by the resources library. The
  resources library is much more efficient than calling
  windower.ffxi.get_player().

On branch macrochanger-update
Changes to be committed:
	modified:   addons/macrochanger/macrochanger.lua

eeach committed:
Update globaldisable to use a bool as opposed to int

Change the name of 'globaldisable' to 'enabled'.

Update this variable to take a boolean as true/false is semantically
  easier to read than 1/0.

Add handling of the case that the user enters some unexpected value.

On branch macrochanger-update
 Changes to be committed:
	modified:   addons/macrochanger/macrochanger.lua

cairthenn committed:
Update macrochanger.lua

eeach committed:
Updated addon_command options

Move options out of the main script.

Treat options passed as key value pairs.

Add warning about config file changes.

Add warning about deprecated addon_command options.

Update help verbage.

On branch macrochanger-update
Changes to be committed:
	modified:   addons/macrochanger/macrochanger.lua
	new file:   addons/macrochanger/options.lua

eeach committed:
Remove out-of-context hanging indents

Remove superfluous spaces from messages in job_change so that they
  look better in the game log.

On branch macrochanger-update
Changes to be committed:
	modified:   addons/macrochanger/macrochanger.lua

eeach committed:
Remove settings.xml

Remove settings.xml as it will be auto-generated on the first run
  with the default settings.

On branch macrochanger-update
Changes to be committed:
	deleted:    data/settings.xml

eeach committed:
Replace <your name> with appropriate data

Replace <your name> in the license block with the names of
  contributors.

On branch macrochanger-update
Changes to be committed:
	modified:   addons/macrochanger/macrochanger.lua

eeach committed:
Add README.md detailing command options and configuration

Add README.md with an explanation of the addon commands available
  and an example of the configuration found in settings.xml.

On branch macrochanger-update
Changes to be committed:
	new file:   addons/macrochanger/README.md

eeach committed:
Add missing '_' in Configuration Example

On branch macrochanger-update
Changes to be committed:
	modified:   addons/macrochanger/README.md

eeach committed:
Add 'help' option

Add an option to addon_command() that displays the help text.

On branch macrochanger-update
Changes to be committed:
	modified:   addons/macrochanger/macrochanger.lua

eeach committed:
Update check for default settings

Use an explicit flag to detect default settings as opposed to a
  cross-typed variable.

On branch macrochanger-update
Changes to be committed:
	modified:   addons/macrochanger/macrochanger.lua

z16 commented:
I've been on vacation for a while, but I'll return home at the end of the weekend, then I'll look over it again. Sorry it's taking so long!

eeach committed:
Fix infinite loop on 'mc help'

Add a return after options.help() to the parameter loop.

On branch macrochanger-update

Changes to be committed:
	modified:   addons/macrochanger/macrochanger.lua
	modified:   addons/macrochanger/options.lua

eeach committed:
Merge branch 'dev' of github.com:Windower/Lua into macrochanger-update

This branch is 18 commits ahead, 66 commits behind Windower:dev.

eeach commented:
Oh, no worries. I thought you were waiting for me.

I had a bunch of IRL work stuff going on, so I'm sorry it took me so long to get back to this.

z16 commented:
Sorry, same here, this completely slipped my mind. I'll try to give it a rundown in a few hours.

z16 commented on addons/macrochanger/options.lua:
```diff
@@ -0,0 +1,47 @@
+options = {}
```
It's cleaner to define a local variable `options` and then return it at the end of the module. That way you don't have to modify global state inside a module and can reuse it easier. (You'd then have to include it by assigning it.)

z16 commented on addons/macrochanger/macrochanger.lua:
```diff
@@ -26,143 +27,67 @@
+            options.help()
+            return
+        end
+    end
```
Is this mechanism meant to loop over multiple options in the same line? While that wouldn't really be an issue, it prevents passing an arbitrary number of arguments to the respective functions (although it seems that this addon in particular doesn't need that). And you don't really gain much from it, since you can easily chain commands with `;`.

z16 commented on addons/macrochanger/README.md:
````diff
@@ -0,0 +1,47 @@
+Disables all automated macro switching.
+
+### Configuration Example: ###
+```
````
If you add `xml` to this line you'll get syntax coloring.

z16 reviewed (approved):
I added some comments, it looks good to me. The few things I mentioned are very much optional. Let me know if you still want to adjust something or if you want me to merge it as it is.
